packages/flutter_tools/lib/src/commands/update_packages.dart

```diff
@@ -259,1548 +169,327 @@ class UpdatePackagesCommand extends FlutterCommand {
-    // pubspec.lock files with the right transitive dependencies.
-    //
-    // This can be expensive, so we run them in parallel. If we hadn't already
-    // warmed the cache above, running them in parallel could be dangerous due
-    // to contention when unpacking downloaded dependencies, but since we have
-    // downloaded all that we need, it is safe to run them in parallel.
-    final Status status = globals.logger.startProgress(
-      'Running "flutter pub get" in affected packages...',
-    );
-    try {
-      // int.tryParse will not accept null, but will convert empty string to null
-      final int? maxJobs = int.tryParse(stringArg('jobs') ?? '');
-      final TaskQueue<void> queue = TaskQueue<void>(maxJobs: maxJobs);
-      for (final Directory dir in packages) {
-        unawaited(
-          queue.add(() async {
-            final Stopwatch stopwatch = Stopwatch();
-            stopwatch.start();
-            await pub.get(
-              context: PubContext.updatePackages,
-              project: FlutterProject.fromDirectory(dir),
-              // All dependencies should already have been downloaded by the fake
-              // package, so the concurrent checks can all happen offline.
-              offline: true,
-              outputMode: PubOutputMode.failuresOnly,
-            );
-            stopwatch.stop();
-            final double seconds = stopwatch.elapsedMilliseconds / 1000.0;
-            final String relativeDir = globals.fs.path.relative(dir.path, from: Cache.flutterRoot);
-            globals.printStatus('Ran pub get in $relativeDir in ${seconds.toStringAsFixed(1)}s...');
-          }),
-        );
-        count += 1;
-      }
-      unawaited(
-        queue.add(() async {
-          final Stopwatch stopwatch = Stopwatch();
-          await _downloadCoverageData();
-          stopwatch.stop();
-          final double seconds = stopwatch.elapsedMilliseconds / 1000.0;
-          globals.printStatus(
-            'Downloaded lcov data for package:flutter in ${seconds.toStringAsFixed(1)}s...',
-          );
-        }),
+  Future<Map<String, Map<String, String>>> _upgrade(
+    bool forceUpgrade,
+    ({String package, String version})? cherryPick,
+    FlutterProject project,
+    bool relaxToAny,
+  ) async {
+    final Map<String, String> pinnedDeps;
+    if (forceUpgrade) {
+      globals.printStatus('Upgrading packages versions...');
+      pinnedDeps = kManuallyPinnedDependencies;
+    } else if (cherryPick != null) {
+      globals.printStatus(
+        'Pinning package "${cherryPick.package}" to version "${cherryPick.version}"...',
       );
-      await queue.tasksComplete;
-      status.stop();
-      // The exception is rethrown, so don't catch only Exceptions.
-    } catch (exception) {
-      status.cancel();
-      rethrow;
+      pinnedDeps = <String, String>{cherryPick.package: cherryPick.version};
+    } else {
+      throw StateError('');
```

Comment:
Can we include more information with this exception? If this can happen due to incorrect user input, we should `throwToolExit(...)`.